File: Week8/5.java
class Solution {
    public String longestPalindrome(String s) {
        int L=s.length();
        int p=(int)1e9+7;
        long[] pre=new long[L+1];
        long[] last=new long[L+1];
        long[] p131=new long[L+1];
        p131[0]=1;
        for(int i=0;i<L;i++){
            pre[i+1]=(pre[i]*131+(s.charAt(i)-'a'+1))%p;
            p131[i+1]=p131[i]*131%p;
        }
        for(int i=L-1;i>=0;i--){
            last[i]=(last[i+1]*131+(s.charAt(i)-'a'+1))%p;
        }
        for(int l=L;l>=1;l--){
            for(int left=0;left<=L-l;left++){
                int right=left+l-1;
                if(((pre[right+1]-pre[left]*p131[l])%p+p)%p==((last[left]-last[right+1]*p131[l])%p+p)%p){
                    return s.substring(left,right+1);
                }
            }
        }
        return s.substring(0,1);
    }
}

File: Week8/5.js
var longestPalindrome = function(s) {
    var n=s.length
    if(n==0){
        return ""
    }
    anslen=1
    ansstart=0
    for(let center=1;center<n-1;center++){
        let l=center-1
        let r=center+1
        while((l>=0) && (r<n) && (s[l]==s[r])){
            l-=1
            r+=1
        }
        if (r-l-1>anslen){
            anslen=r-l-1
            ansstart=l+1
        }
    }
    for(let center=1;center<n;center++){
        let l=center-1
        let r=center
        while(l>=0 && r<n && s[l]==s[r]){
            l-=1
            r+=1
        }
        if (r-l-1>anslen){
            anslen=r-l-1
            ansstart=l+1
        }
    }
    return s.substring(ansstart,ansstart+anslen)
};